src/perlDebug.ts

```diff
@@ -756,6 +770,119 @@ class PerlDebugSession extends LoggingDebugSession {
 				this.sendResponse(response);
 			});
 	}
+
+	private async loadedSourcesRequestAsync(response: DebugProtocol.LoadedSourcesResponse, args: DebugProtocol.LoadedSourcesArguments): Promise<DebugProtocol.LoadedSourcesResponse> {
+
+		const loadedFiles = await this.perlDebugger.getExpressionValue(
+			`join "\t", grep { /^_</ } keys %main::`
+		);
+
+		const newFiles = loadedFiles
+			.split(/\t/)
+			.filter(x => !/^_<\(eval \d+\)/.test(x))
+			.map(x => x.replace(/^_</, ''))
+			.filter(x => !this._loadedSources.has(x));
+
+		for (const file of newFiles) {
+
+			const newSource = new Source(
+				file,
+				file,
+				// no sourceReference when debugging locally, so vscode will
+				// open the local file rather than retrieving a read-only
+				// version of the code through the debugger (that lacks code
+				// past `__END__` markers, among possibly other limitations).
+				this.perlDebugger.isRemote
+					? this._loadedSources.size
+					: 0
+			);
+
+			this.sendEvent(new LoadedSourceEvent("new", newSource));
+
+			this._loadedSources.set(file, newSource);
+
+		}
+
+		response.body = {
+			sources: [...this._loadedSources.values()]
+		};
+
+		return response;
+
+	}
+
+	protected loadedSourcesRequest(response: DebugProtocol.LoadedSourcesResponse, args: DebugProtocol.LoadedSourcesArguments) {
+
+		this.loadedSourcesRequestAsync(response, args)
+			.then(res => this.sendResponse(res))
+			.catch(err => {
+
+				const [ error = err ] = err.errors || [];
+				this.sendEvent(new OutputEvent(`--->Loaded sources request error...${error.message}\n`));
+				response.body = {
+					sources: [
+					]
+				};
+				this.sendResponse(response);
+			});
+
+	}
+
+	private async sourceRequestAsync(response: DebugProtocol.SourceResponse, args: DebugProtocol.SourceArguments): Promise<DebugProtocol.SourceResponse> {
+
+		// NOTE(bh): When sources reported by `loadedSources` have some
+		// non-zero `sourceReference` specified, Visual Studio Code will
+		// ask us for the source code, otherwise it interprets the `path`
+		// as a local file. Our `loadedSources` takes the paths from the
+		// Perl debugger, and `sourceReference` is just a counter value.
+		// Accordingly there is no point for us to distinguish the cases.
+
+		if (args.source && args.source.sourceReference) {
+			// retrieve by source reference
+		} else {
+			// retrieve by path
+		}
+
+		const escapedPath = args.source.path.replace(
+			/([\\'])/g,
+			'\\$1'
+		);
+
+		response.body = {
```

Comment:
We should move this low level stuff into `perlDebuggerConnection` in the `adapter.ts` code - the `perlDebug.ts` is the wireup between vs code and the adapter